In [214]:
import pandas as pd
import os
import numpy as np
from getpass import getuser

In [215]:
# =============================
# CONFIG
# =============================
USER = getuser()
INPUT_DIR = f"C:/Users/{USER}/Documents/GitHub/tennis-homophily/data/atp/"
INPUT_FILE = os.path.join(INPUT_DIR, "men_matches_with_ranks.xlsx")
OUTPUT_FILE = os.path.join(INPUT_DIR, "win_lose_df.xlsx")
CLEANED_FILE = os.path.join(INPUT_DIR, "men_matches_with_ranks_cleaned.xlsx")
gravity_lang_lookup_dir = os.path.join(INPUT_DIR, "gravity_lang_lookup.xlsx")


In [216]:

# =============================
# Load Dataset
# =============================
print("Loading Grand Slam matches…")
if not os.path.exists(INPUT_FILE):
    raise FileNotFoundError(f"Grand Slam file not found: {INPUT_FILE}")

df = pd.read_excel(INPUT_FILE)
print(df.shape)
df.columns.tolist()


Loading Grand Slam matches…
(1405, 252)


['match_id',
 'tournament',
 'location',
 'date',
 'year',
 'tournament_code',
 'stage',
 'match_duration',
 'winners_set1',
 'winners_set2',
 'winners_set3',
 'winners_set1_tiebreak',
 'winners_set2_tiebreak',
 'winners_set3_tiebreak',
 'losers_set1',
 'losers_set2',
 'losers_set3',
 'losers_set1_tiebreak',
 'losers_set2_tiebreak',
 'losers_set3_tiebreak',
 'winners_set4',
 'winners_set5',
 'losers_set4',
 'losers_set5',
 'winners_set4_tiebreak',
 'losers_set4_tiebreak',
 'losers_set5_tiebreak',
 'winners_p1_name',
 'winners_p1_surname',
 'winners_p1_ranking',
 'winners_p1_status',
 'winners_p2_name',
 'winners_p2_surname',
 'winners_p2_ranking',
 'winners_p2_status',
 'losers_p1_name',
 'losers_p1_surname',
 'losers_p1_ranking',
 'losers_p1_status',
 'losers_p2_name',
 'losers_p2_surname',
 'losers_p2_ranking',
 'losers_p2_status',
 'tourn_key',
 'winners_p1_initial',
 'winners_p2_initial',
 'losers_p1_initial',
 'losers_p2_initial',
 'winners_p1_rank_num',
 'winners_p1_dob',
 'winne

# CEPII data language

In [217]:

# path = f"C:/Users/{USER}/Documents/GitHub/tennis-homophily/data/atp/Gravity_V202211.csv"

# # ISO3 presenti nel dataset match
# iso_set = set(df["winners_p1_iso3"]) | \
#           set(df["winners_p2_iso3"]) | \
#           set(df["losers_p1_iso3"]) | \
#           set(df["losers_p2_iso3"])

# usecols = ["year", "iso3_o", "iso3_d", "comlang_off", "comlang_ethno"]

# dtypes = {
#     "year": "int16",
#     "iso3_o": "string",
#     "iso3_d": "string",
#     "comlang_off": "Int8",
#     "comlang_ethno": "Int8",
# }

# chunks = []

# for chunk in pd.read_csv(path, usecols=usecols, dtype=dtypes, chunksize=300_000):

#     # filtro anni
#     chunk = chunk[chunk["year"].between(2018, 2023)]
#     if chunk.empty:
#         continue

#     # filtro paesi rilevanti
#     chunk = chunk[
#         chunk["iso3_o"].isin(iso_set) &
#         chunk["iso3_d"].isin(iso_set)
#     ]

#     if not chunk.empty:
#         chunks.append(chunk)

# gravity_lang = pd.concat(chunks, ignore_index=True)

# gravity_lang.shape


In [218]:
# # Creazione dataset simmetrico (iso3_o, iso3_d) <-> (iso3_d, iso3_o)
# gravity_lang_sym = pd.concat(
#     [
#         gravity_lang,
#         gravity_lang.rename(columns={"iso3_o":"iso3_d","iso3_d":"iso3_o"})
#     ],
#     ignore_index=True
# )

# gravity_lang_sym = gravity_lang_sym.drop_duplicates(
#     subset=["year","iso3_o","iso3_d"]
# )


In [219]:
# # symmetry: same nationality -->  same language
# same = gravity_lang_sym["iso3_o"] == gravity_lang_sym["iso3_d"]
# gravity_lang_sym.loc[same, ["comlang_off","comlang_ethno"]] = 1
# gravity_lang_sym


In [220]:
# # make dataset lookup: (iso3_o, iso3_d) -> (comlang_off, comlang_ethno)
# gravity_lang_lookup = (
#     gravity_lang_sym
#     .groupby(["iso3_o","iso3_d"], as_index=False)[
#         ["comlang_off","comlang_ethno"]
#     ]
#     .max()
# )
# gravity_lang_lookup


In [221]:
# # =============================
# # Save
# # =============================
# print(f"Saving to {gravity_lang_lookup_dir} …")
# with pd.ExcelWriter(gravity_lang_lookup_dir, engine="xlsxwriter") as xlw:
#     gravity_lang_lookup.to_excel(xlw, index=False, sheet_name="main_sheet")

# print("Done.")


In [222]:
gravity_lang_lookup = pd.read_excel(gravity_lang_lookup_dir, sheet_name="main_sheet")

In [223]:
# merge winners
df = df.merge(
    gravity_lang_lookup,
    left_on=["winners_p1_iso3","winners_p2_iso3"],
    right_on=["iso3_o","iso3_d"],
    how="left"
)

df = df.rename(columns={
    "comlang_off":"winners_same_language",
    "comlang_ethno":"winners_linguistic_proximity"
}).drop(columns=["iso3_o","iso3_d"])

# merge losers
df = df.merge(
    gravity_lang_lookup,
    left_on=["losers_p1_iso3","losers_p2_iso3"],
    right_on=["iso3_o","iso3_d"],
    how="left"
)

df = df.rename(columns={
    "comlang_off":"losers_same_language",
    "comlang_ethno":"losers_linguistic_proximity"
}).drop(columns=["iso3_o","iso3_d"])



In [224]:
cols = [
    "winners_same_language",
    "winners_linguistic_proximity",
    "losers_same_language",
    "losers_linguistic_proximity"
]

for c in cols:
    df[c] = df[c].fillna(0).astype(int)


In [225]:
perc_table = (
    df[cols]
    .mean()
    .mul(100)
    .round(2)
    .to_frame(name="Percentage (%)")
)

perc_table


,Percentage (%)
winners_same_language,45.55
winners_linguistic_proximity,46.05
losers_same_language,31.46
losers_linguistic_proximity,32.24


  # Filter dataset

In [226]:
# Rimuovi l'anno 2020 dal dataset, COVID time, Wimblendon non si è giocato
df = df[df['year'] != 2020].copy()

# Controllo rapido
print(df['year'].value_counts())

year
2018    262
2019    250
2021    249
2022    249
2023    238
Name: count, dtype: int64


# Create new variables

In [227]:
# Create surface variable
surface_map = {
    "Australian Open": "Hard",
    "US Open": "Hard",
    "Roland Garros": "Clay",
    "Wimbledon": "Grass"
}

df['surface'] = df['tournament'].map(surface_map)


In [228]:
# Normalize stage names
df['stage_clean'] = df['stage'].str.lower().str.strip()

stage_order = {
    "finals": 7,
    "semi-finals": 6,
    "quarter-finals": 5,
    "round of 16": 4,
    "round of 32": 3,
    "round of 64": 2,
    "2nd round qualifying": 1,
    "1st round qualifying": 0
}

df['stage_code'] = df['stage_clean'].map(stage_order)
df['stage_code'].unique()

array([7, 6, 5, 4, 3, 2, 1, 0], dtype=int64)

In [229]:
cols = [
    "winners_p1_rank_num", "winners_p2_rank_num",
    "losers_p1_rank_num",  "losers_p2_rank_num",
    "winners_p1_singles_career_high_rank_num",
    "winners_p2_singles_career_high_rank_num",
    "losers_p1_singles_career_high_rank_num",
    "losers_p2_singles_career_high_rank_num"
]

for c in cols:
    df[c] = pd.to_numeric(df[c], errors="coerce")

In [230]:
# --- Physical Homophily --- #

# --- hand ---
df["same_hand_winners"] = (df["winners_p1_hand"] == df["winners_p2_hand"]).astype(int)
df["same_hand_losers"]  = (df["losers_p1_hand"]  == df["losers_p2_hand"]).astype(int)

# --- backhand ---
df["same_backhand_winners"] = (df["winners_p1_backhand"] == df["winners_p2_backhand"]).astype(int)
df["same_backhand_losers"]  = (df["losers_p1_backhand"]  == df["losers_p2_backhand"]).astype(int)

# --- height ---
df["height_diff_winners"] = (df["winners_p1_height-cm"] - df["winners_p2_height-cm"]).abs()
df["height_diff_losers"]  = (df["losers_p1_height-cm"]  - df["losers_p2_height-cm"]).abs()

# --- weight ---
df["weight_diff_winners"] = (df["winners_p1_weight-kg"] - df["winners_p2_weight-kg"]).abs()
df["weight_diff_losers"]  = (df["losers_p1_weight-kg"]  - df["losers_p2_weight-kg"]).abs()

# --- experience ---
df["experience_diff_winners"] = (df["winners_p1_experience_double"] - df["winners_p2_experience_double"]).abs()
df["experience_diff_losers"]  = (df["losers_p1_experience_double"]  - df["losers_p2_experience_double"]).abs()

# --- rank / ability ---
df["rank_diff_winners"] = (df["winners_p1_rank_num"] - df["winners_p2_rank_num"]).abs()
df["rank_mean_winners"] = (df["winners_p1_rank_num"] + df["winners_p2_rank_num"]) / 2
df["rank_diff_losers"]  = (df["losers_p1_rank_num"]  - df["losers_p2_rank_num"]).abs()
df["rank_mean_losers"]  = (df["losers_p1_rank_num"]  + df["losers_p2_rank_num"]) / 2

df["single_rank_diff_winners"] = (
    df["winners_p1_singles_career_high_rank_num"] - df["winners_p2_singles_career_high_rank_num"]
).abs()

df["single_rank_diff_losers"] = (
    df["losers_p1_singles_career_high_rank_num"] - df["losers_p2_singles_career_high_rank_num"]
).abs()

In [231]:

# --- Social Homophily --- #

# --- nationality ---
df['same_country_winners'] = (df['winners_p1_country'] == df['winners_p2_country']).astype(int)
df['same_country_losers']  = (df['losers_p1_country']  == df['losers_p2_country']).astype(int)

# --- coach ---
df['same_coach_winners'] = (
    (df['winners_p1_coach'].notna()) &
    (df['winners_p2_coach'].notna()) &
    (df['winners_p1_coach'] == df['winners_p2_coach'])
).astype(int)

df['same_coach_losers'] = (
    (df['losers_p1_coach'].notna()) &
    (df['losers_p2_coach'].notna()) &
    (df['losers_p1_coach'] == df['losers_p2_coach'])
).astype(int)


In [232]:

# Differenza assoluta nel win ratio YTD e Career
df['wl_ytd_diff_winners'] = (df['winners_p1_doubles_win_ratio-ytd'] - df['winners_p2_doubles_win_ratio-ytd']).abs()
df['wl_ytd_diff_losers'] = (df['losers_p1_doubles_win_ratio-ytd']  - df['losers_p2_doubles_win_ratio-ytd']).abs()

df['wl_career_diff_winners'] = (df['winners_p1_doubles_win_ratio-career'] - df['winners_p2_doubles_win_ratio-career']).abs()
df['wl_career_diff_losers']  = (df['losers_p1_doubles_win_ratio-career']  - df['losers_p2_doubles_win_ratio-career']).abs()

# Differenza in titoli (YTD e Career)
df['titles_ytd_diff_winners']    = (df['winners_p1_doubles_titles-ytd'] - df['winners_p2_doubles_titles-ytd']).abs()
df['titles_ytd_diff_losers']     = (df['losers_p1_doubles_titles-ytd'] - df['losers_p2_doubles_titles-ytd']).abs()

df['titles_career_diff_winners'] = (df['winners_p1_doubles_titles-career'] - df['winners_p2_doubles_titles-career']).abs()
df['titles_career_diff_losers']  = (df['losers_p1_doubles_titles-career'] - df['losers_p2_doubles_titles-career']).abs()

# single specialists
df['single_specialist_winners'] = (
  (df['winners_p1_single_specialist'] == 1) & (df['winners_p2_single_specialist'] == 1)
).astype(int)

df['single_specialist_losers']  = (
  (df['losers_p1_single_specialist'] == 1) & (df['losers_p2_single_specialist'] == 1)
).astype(int)

In [233]:
# =============================
# Save
# =============================
print(f"Saving to {CLEANED_FILE} …")
with pd.ExcelWriter(CLEANED_FILE, engine="xlsxwriter") as xlw:
    df.to_excel(xlw, index=False, sheet_name="players_list")

print("Done.")


Saving to C:/Users/ALESSANDRO/Documents/GitHub/tennis-homophily/data/atp/men_matches_with_ranks_cleaned.xlsx …
Done.


 # some stats

In [234]:
# variables for descriptive stats
desc_vars = [
    "same_coach_winners",
    "same_country_winners",
    "winners_same_language",
    "winners_linguistic_proximity",
    "same_hand_winners",
    "same_backhand_winners",
    "height_diff_winners",
    "weight_diff_winners",
    "experience_diff_winners"
]

In [235]:
def descriptive_means(data, variables):
    out = data[variables].mean().reset_index()
    out.columns = ["variable", "mean"]
    return out

## single specialist + best in pair

In [236]:
### WINNERS ###

w_p1_single = df["winners_p1_single_specialist"].eq(1)
w_p2_single = df["winners_p2_single_specialist"].eq(1)

w_p1_best = df["winners_p1_rank_num"] < df["winners_p2_rank_num"]
w_p2_best = df["winners_p2_rank_num"] < df["winners_p1_rank_num"]

df["winner_specialist_best"] = (
    (w_p1_single & w_p1_best) |
    (w_p2_single & w_p2_best)
).astype(int)


### LOSERS ###

l_p1_single = df["losers_p1_single_specialist"].eq(1)
l_p2_single = df["losers_p2_single_specialist"].eq(1)

l_p1_best = df["losers_p1_rank_num"] < df["losers_p2_rank_num"]
l_p2_best = df["losers_p2_rank_num"] < df["losers_p1_rank_num"]

df["loser_specialist_best"] = (
    (l_p1_single & l_p1_best) |
    (l_p2_single & l_p2_best)
).astype(int)

In [237]:
df["structure_diff"] = df["winner_specialist_best"] - df["loser_specialist_best"]

structure_counts = df["structure_diff"].value_counts()
print("Structure comparison:")
print(structure_counts)

advantage_wins = (df["structure_diff"] == 1).sum()
advantage_losses = (df["structure_diff"] == -1).sum()

if (advantage_wins + advantage_losses) > 0:
    advantage_rate = advantage_wins / (advantage_wins + advantage_losses)
else:
    advantage_rate = np.nan

print("\nWin rate when structure differs:", round(advantage_rate,4))


Structure comparison:
structure_diff
 0    1201
-1      25
 1      22
Name: count, dtype: int64

Win rate when structure differs: 0.4681


In [238]:
df_winner_advantage = df[df["structure_diff"] == 1].copy()
df_loser_advantage  = df[df["structure_diff"] == -1].copy()

When teams differ in the presence of a singles specialist who is also the best-ranked doubles
player within the pair, the team with this structure wins 46.8% of the matches (22 out of 47).

## does the team with the best current double ranked player win more often?

In [249]:
# best player in each team
df["winner_best_rank"] = df[["winners_p1_rank_num", "winners_p2_rank_num"]].min(axis=1)
df["loser_best_rank"]  = df[["losers_p1_rank_num", "losers_p2_rank_num"]].min(axis=1)

# remove ties
df_valid = df[df["winner_best_rank"] != df["loser_best_rank"]]

# best-player team wins
best_player_team_wins = (df_valid["winner_best_rank"] < df_valid["loser_best_rank"]).sum()
print("Best-player team wins:", best_player_team_wins)

# best-player team loses
best_player_team_losses = (df_valid["winner_best_rank"] > df_valid["loser_best_rank"]).sum()
print("Best-player team loses:", best_player_team_losses)

# probability best-player team wins
share = best_player_team_wins / (best_player_team_wins + best_player_team_losses)
print("Share of wins for best-player teams:", round(share, 4))



Best-player team wins: 496
Best-player team loses: 273
Share of wins for best-player teams: 0.645


### partner's choice

In [240]:
# subset: matches where the winning team contains the best-ranked player
df_best_player_team_wins = df_valid[
    df_valid["winner_best_rank"] < df_valid["loser_best_rank"]
].copy()

In [241]:
descriptive_means(df_best_player_team_wins, desc_vars)

,variable,mean
0,same_coach_winners,0.092742
1,same_country_winners,0.342742
2,winners_same_language,0.530242
3,winners_linguistic_proximity,0.532258
4,same_hand_winners,0.699597
5,same_backhand_winners,0.637097
6,height_diff_winners,5.650924
7,weight_diff_winners,5.587269
8,experience_diff_winners,5.341410


## does the team with the best single ranked player win more often?

In [242]:
# best single player in each team
df["winner_best_single"] = df[
    ["winners_p1_singles_career_high_rank_num", "winners_p2_singles_career_high_rank_num"]
].min(axis=1)

df["loser_best_single"] = df[
    ["losers_p1_singles_career_high_rank_num", "losers_p2_singles_career_high_rank_num"]
].min(axis=1)

In [243]:
df_valid = df[
    df["winner_best_single"].notna() &
    df["loser_best_single"].notna() &
    (df["winner_best_single"] != df["loser_best_single"])
].copy()

In [251]:
best_single_player_team_wins = (df_valid["winner_best_single"] < df_valid["loser_best_single"]).sum()
best_single_player_team_losses = (df_valid["winner_best_single"] > df_valid["loser_best_single"]).sum()
share = best_single_player_team_wins / (best_single_player_team_wins + best_single_player_team_losses)

print("best_single_player_team_wins:", best_single_player_team_wins)
print("best_single_player_team_losses:", best_single_player_team_losses)
print("total matches:", best_single_player_team_wins + best_single_player_team_losses)
print("winning share:", round(share, 4))

best_single_player_team_wins: 494
best_single_player_team_losses: 466
total matches: 960
winning share: 0.5146


### partner's choice

In [245]:
# subset: matches where the winning team contains the best-ranked player
df_best_single_player_team_wins = df_valid[
    df_valid["winner_best_single"] < df_valid["loser_best_single"]
].copy()

In [246]:
descriptive_means(df_best_single_player_team_wins, desc_vars)

,variable,mean
0,same_coach_winners,0.044534
1,same_country_winners,0.372470
2,winners_same_language,0.544534
3,winners_linguistic_proximity,0.552632
4,same_hand_winners,0.696356
5,same_backhand_winners,0.540486
6,height_diff_winners,6.106904
7,weight_diff_winners,6.527840
8,experience_diff_winners,5.393401


## experience vs ranking 

In [252]:
# ----- WINNERS -----
winner_best_rank_p1 = df["winners_p1_rank_num"] < df["winners_p2_rank_num"]
winner_best_rank_p2 = df["winners_p2_rank_num"] < df["winners_p1_rank_num"]

winner_more_exp_p1 = df["winners_p1_experience_double"] > df["winners_p2_experience_double"]
winner_more_exp_p2 = df["winners_p2_experience_double"] > df["winners_p1_experience_double"]

# cases where best-ranked player is also the more experienced one
winner_best_rank_more_exp = (
    (winner_best_rank_p1 & winner_more_exp_p1) |
    (winner_best_rank_p2 & winner_more_exp_p2)
)

# optional: exclude ties in rank or experience
winner_valid = (
    df["winners_p1_rank_num"].ne(df["winners_p2_rank_num"]) &
    df["winners_p1_experience_double"].ne(df["winners_p2_experience_double"])
)

winner_count = winner_best_rank_more_exp[winner_valid].sum()
winner_total = winner_valid.sum()
winner_share = winner_count / winner_total

print("Winners:")
print("Count =", winner_count)
print("Total valid =", winner_total)
print("Share =", round(winner_share, 4))


# ----- LOSERS -----
loser_best_rank_p1 = df["losers_p1_rank_num"] < df["losers_p2_rank_num"]
loser_best_rank_p2 = df["losers_p2_rank_num"] < df["losers_p1_rank_num"]

loser_more_exp_p1 = df["losers_p1_experience_double"] > df["losers_p2_experience_double"]
loser_more_exp_p2 = df["losers_p2_experience_double"] > df["losers_p1_experience_double"]

# cases where best-ranked player is also the more experienced one
loser_best_rank_more_exp = (
    (loser_best_rank_p1 & loser_more_exp_p1) |
    (loser_best_rank_p2 & loser_more_exp_p2)
)

# optional: exclude ties in rank or experience
loser_valid = (
    df["losers_p1_rank_num"].ne(df["losers_p2_rank_num"]) &
    df["losers_p1_experience_double"].ne(df["losers_p2_experience_double"])
)

loser_count = loser_best_rank_more_exp[loser_valid].sum()
loser_total = loser_valid.sum()
loser_share = loser_count / loser_total

print("\nLosers:")
print("Count =", loser_count)
print("Total valid =", loser_total)
print("Share =", round(loser_share, 4))

Winners:
Count = 365
Total valid = 1088
Share = 0.3355

Losers:
Count = 255
Total valid = 1168
Share = 0.2183
